distributed/worker_memory.py

```diff
@@ -0,0 +1,406 @@
+                    "Process memory: %s -- Worker memory limit: %s",
+                    int(frac * 100),
+                    format_bytes(memory),
+                    format_bytes(self.memory_limit)
+                    if self.memory_limit is not None
+                    else "None",
+                )
+                worker.status = Status.paused
+        elif worker.status == Status.paused:
+            logger.warning(
+                "Worker is at %d%% memory usage. Resuming worker. "
+                "Process memory: %s -- Worker memory limit: %s",
+                int(frac * 100),
+                format_bytes(memory),
+                format_bytes(self.memory_limit)
+                if self.memory_limit is not None
+                else "None",
+            )
+            worker.status = Status.running
+
+    async def _maybe_spill(self, worker: Worker, memory: int) -> None:
+        if self.memory_spill_fraction is False:
+            return
+
+        # SpillBuffer or a duct-type compatible MutableMapping which offers the
+        # fast property and evict() methods. Dask-CUDA uses this.
+        if not hasattr(self.data, "fast") or not hasattr(self.data, "evict"):
+            return
+        data = cast(ManualEvictProto, self.data)
+
+        assert self.memory_limit
+        frac = memory / self.memory_limit
+        if frac <= self.memory_spill_fraction:
+            return
+
+        total_spilled = 0
+        logger.debug(
+            "Worker is at %.0f%% memory usage. Start spilling data to disk.",
+            frac * 100,
+        )
+        # Implement hysteresis cycle where spilling starts at the spill threshold and
+        # stops at the target threshold. Normally that here the target threshold defines
+        # process memory, whereas normally it defines reported managed memory (e.g.
+        # output of sizeof() ). If target=False, disable hysteresis.
+        target = self.memory_limit * (
+            self.memory_target_fraction or self.memory_spill_fraction
+        )
+        count = 0
+        need = memory - target
+        while memory > target:
+            if not data.fast:
+                logger.warning(
+                    "Unmanaged memory use is high. This may indicate a memory leak "
+                    "or the memory may not be released to the OS; see "
+                    "https://distributed.dask.org/en/latest/worker.html#memtrim "
+                    "for more information. "
+                    "-- Unmanaged memory: %s -- Worker memory limit: %s",
+                    format_bytes(memory),
+                    format_bytes(self.memory_limit),
+                )
+                break
+
+            weight = data.evict()
+            if weight == -1:
+                # Failed to evict:
+                # disk full, spill size limit exceeded, or pickle error
+                break
+
+            total_spilled += weight
+            count += 1
+            await asyncio.sleep(0)
```

Comment:
Because spilling gigabytes or tens of gigabytes to disk can take a very long time and you don't want to send the event loop into apnoea for what could possibly be anywhere between ten seconds to a few minutes.
This is not a change from main.

Reply:
See also: https://github.com/dask/distributed/issues/4424